File: frontend/src/app/integrations/submodules/spotify/services/spotify-player.service.ts
import { HttpClient } from '@angular/common/http'
import { Injectable } from '@angular/core'

import { BehaviorSubject } from 'rxjs'

import {
	SpotifyCurrentlyPlaying,
	SpotifyCurrentlyPlayingContext,
} from '../models/spotify-player.model'
import { SpotifyTrack } from '../models/spotify.model'

@Injectable({
	providedIn: 'root',
})
export class SpotifyPlayerService {
	private readonly baseUrl = '/api/v1/integrations/spotify/player'
	private currentTrackSubject = new BehaviorSubject<SpotifyTrack | null>(null)

	currentTrack$ = this.currentTrackSubject.asObservable()

	constructor(private http: HttpClient) {}

	getPlaybackState() {
		return this.http.get<SpotifyCurrentlyPlayingContext>(this.baseUrl)
	}

	setPlaybackVolume(volume: number) {
		return this.http.put<void>(`${this.baseUrl}/volume?volume=${volume}`, {})
	}

	startResumePlayback(trackUri: string) {
		return this.http.post<void>(`${this.baseUrl}/start?trackUri=${trackUri}`, {})
	}

	pausePlayback() {
		return this.http.post<void>(`${this.baseUrl}/pause`, {})
	}

	getCurrentPlayingTrack() {
		return this.http.get<SpotifyCurrentlyPlaying>(`${this.baseUrl}/current-track`)
	}

	skipToNextTrack() {
		return this.http.post<void>(`${this.baseUrl}/next`, {})
	}

	skipToPreviousTrack() {
		return this.http.post<void>(`${this.baseUrl}/previous`, {})
	}

	setTrackProgress(positionMs: number) {
		return this.http.put<void>(`${this.baseUrl}/progress?positionMs=${positionMs}`, {})
	}

	setCurrentTrack(track: SpotifyTrack | null) {
		this.currentTrackSubject.next(track)
	}
}

File: backend/src/main/java/org/kaiteki/backend/integrations/modules/spotify/services/SpotifyPlayerService.java
package org.kaiteki.backend.integrations.modules.spotify.services;


import com.google.gson.JsonArray;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;
import org.springframework.web.server.ResponseStatusException;
import se.michaelthelin.spotify.SpotifyApi;
import se.michaelthelin.spotify.model_objects.miscellaneous.CurrentlyPlaying;
import se.michaelthelin.spotify.model_objects.miscellaneous.CurrentlyPlayingContext;
import se.michaelthelin.spotify.requests.data.player.*;


@Service
@RequiredArgsConstructor
public class SpotifyPlayerService {
    private final SpotifyCredentialsService spotifyCredentialsService;
    private final SpotifyService spotifyService;

    public CurrentlyPlayingContext getPlaybackState() {
        SpotifyApi authSpotifyApi = spotifyCredentialsService.getAuthSpotifyApi(spotifyService.getSpotifyApi());
        GetInformationAboutUsersCurrentPlaybackRequest request = authSpotifyApi.getInformationAboutUsersCurrentPlayback().build();

        try {
            return request.execute();
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to get users current playing context: " + e.getMessage());
        }
    }

    public void startResumePlayback(String trackUri) {
        SpotifyApi authSpotifyApi = spotifyCredentialsService.getAuthSpotifyApi(spotifyService.getSpotifyApi());

        JsonArray trackUris = new JsonArray();
        trackUris.add(trackUri);

        StartResumeUsersPlaybackRequest request = authSpotifyApi.startResumeUsersPlayback().uris(trackUris).build();

        try {
            request.execute();
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to start/resume playback: " + e.getMessage());
        }
    }

    public void pausePlayback() {
        SpotifyApi authSpotifyApi = spotifyCredentialsService.getAuthSpotifyApi(spotifyService.getSpotifyApi());
        PauseUsersPlaybackRequest request = authSpotifyApi.pauseUsersPlayback().build();

        try {
            request.execute();
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to start/resume playback: " + e.getMessage());
        }
    }

    public void setPlaybackVolume(int volume) {
        SpotifyApi authSpotifyApi = spotifyCredentialsService.getAuthSpotifyApi(spotifyService.getSpotifyApi());
        SetVolumeForUsersPlaybackRequest request = authSpotifyApi.setVolumeForUsersPlayback(volume).build();

        try {
            request.execute();
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to start/resume playback: " + e.getMessage());
        }
    }

    public CurrentlyPlaying getCurrentPlayingTrack() {
        SpotifyApi authSpotifyApi = spotifyCredentialsService.getAuthSpotifyApi(spotifyService.getSpotifyApi());
        GetUsersCurrentlyPlayingTrackRequest request = authSpotifyApi.getUsersCurrentlyPlayingTrack().build();

        try {
            return request.execute();
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to start/resume playback: " + e.getMessage());
        }
    }

    public void skipToNextTrack() {
        SpotifyApi authSpotifyApi = spotifyCredentialsService.getAuthSpotifyApi(spotifyService.getSpotifyApi());
        SkipUsersPlaybackToNextTrackRequest request = authSpotifyApi.skipUsersPlaybackToNextTrack().build();

        try {
            request.execute();
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to skip to next track: " + e.getMessage());
        }
    }

    public void skipToPreviousTrack() {
        SpotifyApi authSpotifyApi = spotifyCredentialsService.getAuthSpotifyApi(spotifyService.getSpotifyApi());
        SkipUsersPlaybackToPreviousTrackRequest request = authSpotifyApi.skipUsersPlaybackToPreviousTrack().build();

        try {
            request.execute();
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to skip to next track: " + e.getMessage());
        }
    }


    public void setTrackProgress(int position_ms) {
        SpotifyApi authSpotifyApi = spotifyCredentialsService.getAuthSpotifyApi(spotifyService.getSpotifyApi());
        SeekToPositionInCurrentlyPlayingTrackRequest request = authSpotifyApi.seekToPositionInCurrentlyPlayingTrack(position_ms).build();

        try {
            request.execute();
        } catch (Exception e) {
            throw new ResponseStatusException(HttpStatus.INTERNAL_SERVER_ERROR, "Failed to skip to next track: " + e.getMessage());
        }
    }
}
